kernel/src/process.rs

```diff
@@ -1605,4 +1531,100 @@ impl<C: 'static + Chip> Process<'a, C> {
             }
         });
     }
+
+    fn restart_app(&self, restart_policy: &'static dyn ProcessRestartPolicy) {
+        {
+            // Start with the generic terminate operations. This frees state
+            // for this process and removes any pending tasks from the
+            // scheduler's queue.
+            self.terminate(State::Exited);
+
+            // Decide what to do with this process. Should it be restarted?
+            // Or should we leave it in a stopped & faulted state? If the
+            // process is faulting too often we might not want to restart.
+            // If we are not going to restart the process then we can just
+            // leave it in the stopped faulted state by returning
+            // immediately. This has the same effect as using the
+            // `FaultResponse::Stop` policy.
+            if !restart_policy.should_restart(self) {
+                return;
+            }
```

Comment:
If `restart_app()` is called in response to a fault and then `restart_policy.should_restart()` indicates that the app should not in fact restart, the process will now be left in the `State::Exited` state when before it would have been left in the `State::StoppedFaulted` state. To prevent this bug, I think `restart_app()` needs to take in the state that the process should be left in if restarting fails.